substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/traceid/JfrTraceIdLoadBarrier.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (c) 2020, 2020, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.oracle.svm.jfr.traceid;
+
+import java.util.LinkedList;
+import java.util.Queue;
+import java.util.function.Consumer;
+
+public class JfrTraceIdLoadBarrier {
+    private static Queue<Class<?>> klassQueueOne = null;
```

Comment:
As this infrastructure must be uninterruptible, I don't think that will be possible to use one shared data structure for all objects that need a trace id. Instead, it is probably necessary to use a slightly different approach for each kind of object, as I sketched in the subclasses of `JfrRepository`.

Reply:
As uninterruptible code can't allocate any objects on the Java heap, it will be necessary to use one of the following approaches for the data structures:
a. use native memory: this would work for most use cases but you can't store references to arbitrary Java heap objects in native memory (the GC wouldn't know when to update the reference). However, you can store references to image heap objects, e.g., all `Class` objects, as those are never garbage collected/moved.
b. allocate fixed size data structures at image build time: this is good for startup performance but will increase the image size. Only feasible if the complete set of data is known during the image build. Would for example be feasible for the types as the number of types is known at image build time.
c. using a mix of native and image heap data structures: I for example used that in `UninterruptibleHashtable`.

Reply:
Thanks, Christian, that is helpful.
I guess another option would be to do it like Hotspot JFR does, and store the trace-ID directly in the corresponding native objects for classes, packages, modules, class-loaders & threads. I am not sure how feasible that would be in SubstrateVM?

Reply:
That is not really feasible for Native Image as most of that data is in the read-only part of the image heap (i.e., this part is really mapped as read-only into memory).

Reply:
Ok. Trace-IDs are only used for: classes, packages, modules, class-loaders and threads (at least, this is the case in Hotspot's JFR). The first 4 are all known and fixed at build-time (right?), so we could use approach a or b for them. (I would like to come up with a compact kind of hashtable that maps those image-heap objects directly to longs).

Threads are only created at run-time, so we have to do something different there. But since threads are not in the image heap, we *could* actually store the trace-ID in the thread itself (or a thread-local field), right?

Reply:
One more question: if I create the table in the image-heap, would it be mapped read-only too? Can I influence that? Because I *do* need the table read-write.

Reply:
Actually, thinking a bit more about it, what we need is a compact and dense table. Ideally, each class, package, module and class-loader would have its unique ID, which would be used to look-up the table entry. This could be the ID from metadata.xml (but doesn't have to). And this could be read-only and thus embedded in the image-heap structures. This ID would then be used directly to look up the trace-ID (with all its overloaded used-, serialized-, epoch-, etc flags). If you agree with this, can you give me a hint where I would find the relevant VM-side structures for those types of objects?

Reply:
Ok, I think I found it. Class is mapped in DynamicHub, and the other objects via Target_* substitutions. DynamicHub already has a typeID, maybe this can be used.

Reply:
All classes, packages, modules and class-loaders are known at image build time. I think that it is possible to allocate further class loaders at run-time but they can't be used to load any classes, so they shouldn't be relevant for JFR.

Regarding threads: yes a thread-local field is the best way. I already introduce one, see `JfrThreadLocal.traceId`. Feel free to adapt that as needed.

Regarding read-only vs. writable image heap: the image heap consists of read-only and writable partitions. At image build time, the static analysis figures out which objects need to be placed in the writable partition and which objects can be placed in the read-only partition (so, if the static analysis sees that you modify an image heap object, it will place it in a writeable partition). There are a couple of VM-internal classes where multiple Native Image parts assume that they are always placed in the read-only part of the (e.g., classes).

The classes already have a `typeId` (see `DynamicHub.typeId`). For packages, modules, and class loaders, Native Image uses the original JDK data structures (with a few substitutions though, see the corresponding `Target_*` classes, e.g. `Target_java_lang_Package`. So, for those classes it should be possible to either inject an `id` or `traceId` field that gets initialized at image build time (see `@Inject` and `@RecomputeFieldValue`).

Reply:
Thanks, Christian, that is very helpful. I think I have everything that I need to know now!

Reply:
Ok one more question: I think I know how to inject the field into e.g, ClassLoader. However, how do I access it? E.g. in JfrFeature.afterAnalysis(), given a ClassLoader object, how do I set Target_java_lang_ClassLoader.jfrID field?

Reply:
You cast to the `ClassLoader` to `Target_java_lang_ClassLoader` using `SubstrateUtil.cast(...)`

Reply:
Ok. This must use the Kind.Custom declClass=Blah.class mechanism, is that right?
How can I get the DynamicHub for a given class at image-build-time (during afterAnalysis())? DynamicHub.fromClass() complains that it cannot cast.

Reply:
Hmm, I don't really need the hub, but I need the typeId that will be stored in the hub. Is there any way I can get to this information?

Reply:
Why do you need the `DynamicHub` or the `typeId` at image build time? Isn't it sufficient to just have the maximum type id (see `DynamicHubSupport.getMaxTypeId()`, which is accessible in `beforeCompilation` or during later image build phases)? 

Accessing the `DynamicHub` is possible but not very convenient: it is possible to fetch the `DynamicHub` via `((SVMHost) analysisUniverse.hostVM()).dynamicHub(analysisType);`. `analysisUniverse` is available via `AfterAnalysisAccess` and the `analysisType` is available via the `MetaAccessProvider` from `AfterAnalysisAccess`.

Reply:
I need to set-up the initial trace-ID map. I can fetch the traceID for each Class from the host VM (via its JFR impl). I need to know the type-ID (the one that ends up in the DynamicHub later) for each Class as the index into the trace-ID map, to store the initial trace-ID there. I believe the typeIDs are only established on beforeCompilation(), and I could get all types via CompilationAccessImpl.getTypes() and from each SharedType, get the DynamicHub via getHub(), and from there establish the typeID->Class->traceID mapping. Does that sound correct?

Reply:
Yes, that sounds good but the mapping only needs to be from `typeID` to `traceID` (`Class` and `DynamicHub` are the same object at runtime). I kind of expected that no extra information from the class would be necessary for creating the clean state but I might be wrong about that.

When a data structure that ends up in the image heap is initialized after static analysis, then it is necessary to annotate the field from which the data structure is reachable with `@UnknownObjectField` or `@UnknownPrimitiveField`. Otherwise, the static analysis would for example figure out that the field is always `null` or `0`.

Reply:
> Yes, that sounds good but the mapping only needs to be from `typeID` to `traceID` (`Class` and `DynamicHub` are the same object at runtime). I kind of expected that no extra information from the class would be necessary for creating the clean state but I might be wrong about that.

Yes, the mapping is typeID->traceID, but in order to find the traceID at build-time I need the corresponding Class.

> When a data structure that ends up in the image heap is initialized after static analysis, then it is necessary to annotate the field from which the data structure is reachable with `@UnknownObjectField` or `@UnknownPrimitiveField`. Otherwise, the static analysis would for example figure out that the field is always `null` or `0`.

Ok, good to know. I think I might need that :-)

Reply:
Alright, I think I'm almost there. I can populate the whole traceID array with classes, packages, modules and classLoaders beforeCompilation. Unfortunately the table doesn't seem to make it into the final image. I have:
JfrRuntimeAccessImpl which is added to ImageSingletons in afterRegistration. It has a field JfrTraceIdMap jfrTraceIdMap which is initially null, I annotated that with @UnknownObjectField. JfrTraceIdMap has a long[] map field, and I also annotated that. The traceIdMap field is set to the actual map in beforeCompilation(). At run-time, the traceIdMap is null. What am I doing wrong?

BTW: I could init the map and arrays earlier, e.g. afterAnalysis, but I don't know yet how many elements I need. Iterating all reachable types in afterAnalysis seems to give a smaller number of types than what I see in beforeCompilation.

Reply:
I pushed my current WIP. A few things seem strange:
- The number of 'reachable' types after analysis is much less than the number of types that I see before compilation. In-fact, the number of types before compilation almost matches the number of all (incl. not-reachable) types after analysis. Maybe I'm doing something wrong there?
- Even then, java.lang.invoke.LambdaForm.BasicType is missing in the list of types after analysis, but shows up before compilation.
- Even though I initialize the trace-ID-map after analysis, and I annotated all relevant fields with @UnknownObjectField, they are still null at run-time.

@christianhaeubl Can you please check what I did, and see what's wrong?